File: sdks/swift/Sources/AstroSDK/OpenBanking/OpenBankingClient.swift
import Foundation

public final class OpenBankingClient {
    private let http: AstroHTTPClient
    init(http: AstroHTTPClient) { self.http = http }

    public func createConsent(
        bankHandle: String,
        scopes: [String],
        redirectUri: String,
        state: String,
        codeChallenge: String,
        codeChallengeMethod: String = "S256"
    ) async throws -> Consent {
        let body = CreateConsentBody(
            bankHandle: bankHandle, scopes: scopes, redirectUri: redirectUri,
            state: state, codeChallenge: codeChallenge, codeChallengeMethod: codeChallengeMethod
        )
        return try await http.perform(http.request(http.url("/ob/consents"), method: "POST", body: body))
    }

    public func getConsent(_ consentId: String) async throws -> Consent {
        try await http.perform(http.request(http.url("/ob/consents/\(consentId)")))
    }

    public func exchangeCode(
        code: String,
        redirectUri: String,
        consentId: String,
        codeVerifier: String
    ) async throws -> TokenResponse {
        let body = TokenBody(
            grantType: "authorization_code",
            code: code,
            redirectUri: redirectUri,
            consentId: consentId,
            codeVerifier: codeVerifier
        )
        return try await http.perform(http.request(http.url("/ob/token"), method: "POST", body: body))
    }

    public func refreshToken(_ refreshToken: String) async throws -> TokenResponse {
        let body = RefreshBody(grantType: "refresh_token", refreshToken: refreshToken)
        return try await http.perform(http.request(http.url("/ob/token"), method: "POST", body: body))
    }

    public func getAccounts(accessToken: String, consentId: String) async throws -> [OBAccount] {
        var req = http.request(
            http.url("/ob/accounts"),
            extraHeaders: ["Authorization": "Bearer \(accessToken)", "X-Consent-Id": consentId]
        )
        let result = try await http.perform(req) as OBAccountsWrapper
        return result.accounts
    }

    public func getTransactions(
        accessToken: String,
        consentId: String,
        accountId: String,
        fromDate: String? = nil,
        toDate: String? = nil
    ) async throws -> [OBTransaction] {
        let url = http.url("/ob/accounts/\(accountId)/transactions", query: ["from_date": fromDate, "to_date": toDate])
        let req = http.request(url, extraHeaders: ["Authorization": "Bearer \(accessToken)", "X-Consent-Id": consentId])
        let result = try await http.perform(req) as OBTransactionsWrapper
        return result.transactions
    }
}

private struct CreateConsentBody: Encodable {
    let bankHandle: String; let scopes: [String]; let redirectUri: String
    let state: String; let codeChallenge: String; let codeChallengeMethod: String
    enum CodingKeys: String, CodingKey {
        case bankHandle = "bank_handle"; case scopes; case redirectUri = "redirect_uri"
        case state; case codeChallenge = "code_challenge"; case codeChallengeMethod = "code_challenge_method"
    }
}

private struct TokenBody: Encodable {
    let grantType: String; let code: String; let redirectUri: String
    let consentId: String; let codeVerifier: String
    enum CodingKeys: String, CodingKey {
        case grantType = "grant_type"; case code; case redirectUri = "redirect_uri"
        case consentId = "consent_id"; case codeVerifier = "code_verifier"
    }
}

private struct RefreshBody: Encodable {
    let grantType: String; let refreshToken: String
    enum CodingKeys: String, CodingKey {
        case grantType = "grant_type"; case refreshToken = "refresh_token"
    }
}

private struct OBAccountsWrapper: Decodable { let accounts: [OBAccount] }
private struct OBTransactionsWrapper: Decodable { let transactions: [OBTransaction] }

File: sdks/flutter/lib/src/openbanking/openbanking_client.dart
import '../astro_http.dart';
import '../models/models.dart';

class OpenBankingClient {
  final AstroHttpClient _http;
  const OpenBankingClient(this._http);

  Future<Consent> createConsent({
    required String bankHandle,
    required List<String> scopes,
    required String redirectUri,
    required String state,
    required String codeChallenge,
    String codeChallengeMethod = 'S256',
    int? expiryDays,
  }) async {
    final body = <String, dynamic>{
      'bank_handle': bankHandle,
      'scopes': scopes,
      'redirect_uri': redirectUri,
      'state': state,
      'code_challenge': codeChallenge,
      'code_challenge_method': codeChallengeMethod,
      if (expiryDays != null) 'expiry_days': expiryDays,
    };
    final json = await _http.post('/ob/consents', body);
    return Consent.fromJson(json as Map<String, dynamic>);
  }

  Future<Consent> getConsent(String consentId) async {
    final json = await _http.get('/ob/consents/$consentId');
    return Consent.fromJson(json as Map<String, dynamic>);
  }

  Future<void> revokeConsent(String consentId) async {
    await _http.delete('/ob/consents/$consentId');
  }

  Future<TokenResponse> exchangeCode({
    required String code,
    required String redirectUri,
    required String consentId,
    required String codeVerifier,
  }) async {
    final json = await _http.post('/ob/token', {
      'grant_type': 'authorization_code',
      'code': code,
      'redirect_uri': redirectUri,
      'consent_id': consentId,
      'code_verifier': codeVerifier,
    });
    return TokenResponse.fromJson(json as Map<String, dynamic>);
  }

  Future<TokenResponse> refreshToken(String refreshToken) async {
    final json = await _http.post('/ob/token', {
      'grant_type': 'refresh_token',
      'refresh_token': refreshToken,
    });
    return TokenResponse.fromJson(json as Map<String, dynamic>);
  }

  Future<List<OBAccount>> getAccounts(String accessToken, String consentId) async {
    final json = await _http.get(
      '/ob/accounts',
      extraHeaders: {
        'Authorization': 'Bearer $accessToken',
        'X-Consent-Id': consentId,
      },
    );
    final list = (json as Map<String, dynamic>)['accounts'] as List;
    return list.map((e) => OBAccount.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<List<OBTransaction>> getTransactions(
    String accessToken,
    String consentId,
    String accountId, {
    String? fromDate,
    String? toDate,
  }) async {
    final json = await _http.get(
      '/ob/accounts/$accountId/transactions',
      params: {'from_date': fromDate, 'to_date': toDate},
      extraHeaders: {
        'Authorization': 'Bearer $accessToken',
        'X-Consent-Id': consentId,
      },
    );
    final list = (json as Map<String, dynamic>)['transactions'] as List;
    return list.map((e) => OBTransaction.fromJson(e as Map<String, dynamic>)).toList();
  }
}
